In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)


import os
print(os.listdir("C:/Users/akhil/Music/MH/Participants_Data_Used_Cars/data"))

data_path = "C:/Users/akhil/Music/MH/Participants_Data_Used_Cars/data"

['.DS_Store', 'Data_Test.xlsx', 'Data_Train.xlsx', 'Sample_submission.xlsx']


#  Load the traing data from the `Data_Train.xlsx` excel file

In [2]:
#Train data
train_data_path = data_path+'/Data_Train.xlsx'
train_full_data = pd.read_excel(train_data_path)

print('****************Train Data:',train_full_data.shape)
print(train_full_data.isnull().sum())
train_full_data.head()


****************Train Data: (6019, 13)
Name                    0
Location                0
Year                    0
Kilometers_Driven       0
Fuel_Type               0
Transmission            0
Owner_Type              0
Mileage                 2
Engine                 36
Power                  36
Seats                  42
New_Price            5195
Price                   0
dtype: int64


Name    Location  Year  Kilometers_Driven  \
0            Maruti Wagon R LXI CNG      Mumbai  2010              72000   
1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
2                      Honda Jazz V     Chennai  2011              46000   
3                 Maruti Ertiga VDI     Chennai  2012              87000   
4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   

  Fuel_Type Transmission Owner_Type     Mileage   Engine      Power  Seats  \
0       CNG       Manual      First  26.6 km/kg   998 CC  58.16 bhp    5.0   
1    Diesel       Manual      First  19.67 kmpl  1582 CC  126.2 bhp    5.0   
2    Petrol       Manual      First   18.2 kmpl  1199 CC   88.7 bhp    5.0   
3    Diesel       Manual      First  20.77 kmpl  1248 CC  88.76 bhp    7.0   
4    Diesel    Automatic     Second   15.2 kmpl  1968 CC  140.8 bhp    5.0   

   New_Price  Price  
0        NaN   1.75  
1        NaN  12.50  
2  8.61 Lakh   4.50  
3        NaN   6.00  
4        NaN  17.74

# Load the traing data from the `Data_Test.xlsx` excel file

In [3]:
#Test data
test_data_path = data_path + '/Data_Test.xlsx'
test_full_data = pd.read_excel(test_data_path)

print('Test Data:\n',test_full_data.shape)
print(test_full_data.isnull().sum())
test_full_data.head()

Test Data:
 (1234, 12)
Name                    0
Location                0
Year                    0
Kilometers_Driven       0
Fuel_Type               0
Transmission            0
Owner_Type              0
Mileage                 0
Engine                 10
Power                  10
Seats                  11
New_Price            1052
dtype: int64


Name    Location  Year  \
0                    Maruti Alto K10 LXI CNG       Delhi  2014   
1              Maruti Alto 800 2016-2019 LXI  Coimbatore  2013   
2  Toyota Innova Crysta Touring Sport 2.4 MT      Mumbai  2017   
3                       Toyota Etios Liva GD   Hyderabad  2012   
4                          Hyundai i20 Magna      Mumbai  2014   

   Kilometers_Driven Fuel_Type Transmission Owner_Type      Mileage   Engine  \
0              40929       CNG       Manual      First  32.26 km/kg   998 CC   
1              54493    Petrol       Manual     Second    24.7 kmpl   796 CC   
2              34000    Diesel       Manual      First   13.68 kmpl  2393 CC   
3             139000    Diesel       Manual      First   23.59 kmpl  1364 CC   
4              29000    Petrol       Manual      First    18.5 kmpl  1197 CC   

       Power  Seats   New_Price  
0   58.2 bhp    4.0         NaN  
1   47.3 bhp    5.0         NaN  
2  147.8 bhp    7.0  25.27 Lakh  
3   null bhp    5.0         NaN  
4  82.85 bhp    5.0         NaN

# Print the unique values in the each column 

In [4]:
import string as str
import datetime as dt

curr_year = dt.datetime.now().year

train_full_data['Name'] = train_full_data['Name'].str.upper()
print('Number of unique cars :',train_full_data['Name'].nunique())
print('----------------------------------------------------------')
print('Unique cities :',train_full_data['Location'].unique())

print('----------------------------------------------------------')
print('Fuel_Types :',train_full_data['Fuel_Type'].unique())

print('----------------------------------------------------------')
print('Owner_Types :',train_full_data['Owner_Type'].unique())

print('----------------------------------------------------------')
print('Transmissions :',train_full_data['Transmission'].unique())

print('----------------------------------------------------------')
print('Seats :',train_full_data['Seats'].unique())

print('----------------------------------------------------------')



train_full_data['Mileage'] = train_full_data['Mileage'].str.replace("/", "")
print(train_full_data['Mileage'].str[-4:].unique())
train_full_data['Age'] = curr_year - train_full_data['Year']


# Similar thing for Test data
test_full_data['Mileage'] = test_full_data['Mileage'].str.replace("/", "")
test_full_data['Name'] = test_full_data['Name'].str.upper()
test_full_data['Age'] = curr_year - test_full_data['Year']

Number of unique cars : 1852
----------------------------------------------------------
Unique cities : ['Mumbai' 'Pune' 'Chennai' 'Coimbatore' 'Hyderabad' 'Jaipur' 'Kochi'
 'Kolkata' 'Delhi' 'Bangalore' 'Ahmedabad']
----------------------------------------------------------
Fuel_Types : ['CNG' 'Diesel' 'Petrol' 'LPG' 'Electric']
----------------------------------------------------------
Owner_Types : ['First' 'Second' 'Fourth & Above' 'Third']
----------------------------------------------------------
Transmissions : ['Manual' 'Automatic']
----------------------------------------------------------
Seats : [ 5.  7.  8.  4.  6.  2. nan 10.  9.  0.]
----------------------------------------------------------
['kmkg' 'kmpl' nan]


# Converting object columns into numerical columns
> `Mileage`, `Engine`, `Power` Columns have its units like `kmkg`, `kmkg`,`CC`, `bhp`.
> remove the units and convert all the columns into numeric columns

# Mileage and Kilometers_Driven

In [5]:
#train_full_data['Kilometers_Driven'] = train_full_data['Kilometers_Driven']/100
train_full_data['Price'] = train_full_data['Price']

In [6]:
#Train Data
train_full_data['Mileage'] = train_full_data['Mileage'].str.replace(" kmkg", "")
train_full_data['Mileage'] = train_full_data['Mileage'].str.replace(" kmpl", "")
train_full_data['Mileage'] = train_full_data['Mileage'].astype(float)


#Test data 

test_full_data['Mileage'] = test_full_data['Mileage'].str.replace(" kmkg", "")
test_full_data['Mileage'] = test_full_data['Mileage'].str.replace(" kmpl", "")
test_full_data['Mileage'] = test_full_data['Mileage'].astype(float)

# Print unique units of Engine and Power

In [7]:
print('Unique Engine units',train_full_data['Engine'].str[-2:].unique())
print('Unique Power units',train_full_data['Power'].str[-3:].unique())

Unique Engine units ['CC' nan]
Unique Power units ['bhp' nan]


# Engine

In [8]:
train_full_data['Engine'] = train_full_data['Engine'].str.replace(' CC','')
train_full_data['Engine'] = train_full_data['Engine'].astype(float)


#test_full_data

test_full_data['Engine'] = test_full_data['Engine'].str.replace(' CC','')
test_full_data['Engine'] = test_full_data['Engine'].astype(float)



# Power

In [9]:
train_full_data['Power'] = train_full_data['Power'].str.replace(' bhp','')
train_full_data['Power'] = train_full_data['Power'].str.replace('null','NaN')
train_full_data['Power'] = train_full_data['Power'].astype(float)


# Test data

test_full_data['Power'] = test_full_data['Power'].str.replace(' bhp','')
test_full_data['Power'] = test_full_data['Power'].str.replace('null','NaN')
test_full_data['Power'] = test_full_data['Power'].astype(float)


# Remove the column `New_Price` from the data set
> Since 5195 records have missing values in the `New_Price`, it would be better to drop the column

In [10]:
#Train
train_full_data.drop('New_Price', axis = 1, inplace = True)

#test
test_full_data.drop('New_Price', axis = 1, inplace = True)

# Drop the column `Year`. Since we have calculated the `Age` of the car

In [11]:
#Train
train_full_data.drop('Year',axis = 1,inplace= True )

#test
test_full_data.drop('Year',axis = 1,inplace= True )

# Create Two new features namely Model and Brand

In [12]:
#Train
train_brand = train_full_data.Name.str.split().str[:3]
train_full_data ['Model'] = train_brand.str[1] + ' '+ train_brand.str[2] 
train_full_data ['Brand'] = train_brand.str[0]

#Test
test_brand = test_full_data.Name.str.split().str[:3]
test_full_data ['Model'] = test_brand.str[1] + ' '+ test_brand.str[2]
test_full_data ['Brand'] = test_brand.str[0]
test_full_data.loc[test_full_data.Name.isna().index,'Model'] = '1.4GSI'

# Replace all `0` values with `np.nan`

In [13]:
#Train
train_full_data['Mileage'].fillna(value=np.NaN, inplace=True)
test_full_data['Mileage'].fillna(value=np.NaN, inplace=True)
train_full_data.loc[train_full_data.Mileage==0,'Mileage']= np.nan
test_full_data.loc[test_full_data.Mileage==0,'Mileage']= np.nan

#Test
train_full_data['Seats'].fillna(value=np.NaN, inplace=True)
test_full_data['Seats'].fillna(value=np.NaN, inplace=True)
train_full_data.loc[train_full_data.Seats==0,'Seats']= np.nan
test_full_data.loc[test_full_data.Seats==0,'Seats']= np.nan


# Get the list of all the models where `Power` or `Engine` or `Seats` or `Mileage` is missing

In [14]:
missing_values_brands_train = list ( train_full_data[(train_full_data.Power.isnull()) | 
                                            (train_full_data.Engine.isnull()) |
                                            (train_full_data.Seats.isnull()) |
                                            (train_full_data.Mileage.isnull())
                                           ].Brand.unique()
                          )


missing_values_brands_test = list ( test_full_data[(test_full_data.Power.isnull()) | 
                                            (test_full_data.Engine.isnull()) |
                                            (test_full_data.Seats.isnull()) |
                                            (test_full_data.Mileage.isnull())
                                           ].Brand.unique()
                          )


missing_values_brands = list (set(missing_values_brands_test) | set(missing_values_brands_train))


# Create a dataframe using Train and Test dataset which has the missing values of all the models which are obtained in `missing_values_brands`

# And create a lookup dataframe for the same

# Obtain `missing_power_model`/`missing_power_brand`

In [15]:
#Train
missing_power_model_train = train_full_data.loc[(  (train_full_data.Power.notna()) 
                                                   & 
                                                   (train_full_data.Brand.isin(missing_values_brands)
                                                   )
                                                ) , 
                                                  ["Brand","Model","Power"]
                                               ]

#Test
missing_power_model_test =  test_full_data.loc[(  (test_full_data.Power.notna()) 
                                                   & 
                                                   (test_full_data.Brand.isin(missing_values_brands)
                                                   )
                                                ) , 
                                                  ["Brand","Model","Power"]
                                               ]




                                        
missing_power_model_all = pd.concat([missing_power_model_train,missing_power_model_test])

missing_power_model =  missing_power_model_all.loc[:,["Power"]].groupby(missing_power_model_all.Model).median()
missing_power_brand =  missing_power_model_all.loc[:,["Power"]].groupby(missing_power_model_all.Brand).mean()

# Obtain `missing_engine_model`/`missing_engine_brand`

In [16]:
#Train
missing_engine_model_train = train_full_data.loc[(  (train_full_data.Engine.notna()) 
                                                   & 
                                                   (train_full_data.Brand.isin(missing_values_brands)
                                                   )
                                                ) , 
                                                  ["Brand","Model","Engine"]
                                               ]

#Test
missing_engine_model_test =  test_full_data.loc[(  (test_full_data.Engine.notna()) 
                                                   & 
                                                   (test_full_data.Brand.isin(missing_values_brands)
                                                   )
                                                ) , 
                                                  ["Brand","Model","Engine"]
                                               ]

             
missing_engine_model_all = pd.concat([missing_engine_model_train,missing_engine_model_test])

missing_engine_model =  missing_engine_model_all.loc[:,["Engine"]].groupby(missing_engine_model_all.Model).median()

missing_engine_brand =  missing_engine_model_all.loc[:,["Engine"]].groupby(missing_engine_model_all.Brand).mean()


# Obtain `missing_seats_model`/`missing_seats_brand`

In [17]:
#Train
missing_Seats_model_train = train_full_data.loc[(  (train_full_data.Seats.notna()) 
                                                   & 
                                                   (train_full_data.Brand.isin(missing_values_brands)
                                                   )
                                                ) , 
                                                  ["Brand","Model","Seats"]
                                               ]

#Test
missing_Seats_model_test =  test_full_data.loc[(  (test_full_data.Seats.notna()) 
                                                   & 
                                                   (test_full_data.Brand.isin(missing_values_brands)
                                                   )
                                                ) , 
                                                  ["Brand","Model","Seats"]
                                               ]

             
missing_Seats_model_all = pd.concat([missing_Seats_model_train,missing_Seats_model_test])

missing_seats_model =  missing_Seats_model_all.loc[:,["Seats"]].groupby(missing_Seats_model_all.Model).median()

missing_seats_brand =  missing_Seats_model_all.loc[:,["Seats"]].groupby(missing_Seats_model_all.Brand).mean()

# Obtain `missing_mileage_model`/`missing_mileage_brand`

In [18]:
#Train
missing_Mileage_model_train = train_full_data.loc[(  (train_full_data.Mileage.notna()) 
                                                   & 
                                                   (train_full_data.Brand.isin(missing_values_brands)
                                                   )
                                                ) , 
                                                  ["Brand","Model","Mileage"]
                                               ]

#Test
missing_Mileage_model_test =  test_full_data.loc[(  (test_full_data.Mileage.notna()) 
                                                   & 
                                                   (test_full_data.Brand.isin(missing_values_brands)
                                                   )
                                                ) , 
                                                  ["Brand","Model","Mileage"]
                                               ]

             
missing_Mileage_model_all = pd.concat([missing_Mileage_model_train,missing_Mileage_model_test])

missing_mileage_model =  missing_Mileage_model_all.loc[:,["Mileage"]].groupby(missing_Mileage_model_all.Model).median()

missing_mileage_brand =  missing_Mileage_model_all.loc[:,["Mileage"]].groupby(missing_Mileage_model_all.Brand).mean()

# Create a methods for each above features to fill the missing values with the values obtained by lookup dataframe

# Method for `Power`

In [19]:
def fill_na_power(df):
    
    if df['Power'] == 0.0:
        if df['Power'] == 0.0 and df['Model'] in missing_power_model.index:
            return missing_power_model.at[df['Model'],'Power']
        elif df.Power == 0.0 and df['Brand'] in missing_power_brand.index:
            return missing_power_brand.at[df['Brand'],'Power']
        else:
            return df['Power']
    else:
        return df['Power']   

# Method for `Engine`

In [20]:
def fill_na_engine(df):
    
    if df['Engine'] == 0.0:
        if df['Engine'] == 0.0 and df['Model'] in missing_engine_model.index:
            return missing_engine_model.at[df['Model'],'Engine']
        elif df.Engine == 0.0 and df['Brand'] in missing_engine_brand.index:
            return missing_engine_brand.at[df['Brand'],'Engine']
        else:
            return df['Engine']
    else:
        return df['Engine']   

# Method for `Seats`

In [21]:
def fill_na_seats(df):
    
    if df['Seats'] == 0.0:
        if df['Seats'] == 0.0 and df['Model'] in missing_seats_model.index:
            return missing_seats_model.at[df['Model'],'Seats']
        elif df.Seats == 0.0 and df['Brand'] in missing_seats_brand.index:
            return missing_seats_brand.at[df['Brand'],'Seats']
        else:
            return df['Seats']
    else:
        return df['Seats']   

# Method for `Mileage`

In [22]:
def fill_na_mileage(df):
    
    if df['Mileage'] == 0.0:
        if df['Mileage'] == 0.0 and df['Model'] in missing_mileage_model.index:
            return missing_mileage_model.at[df['Model'],'Mileage']
        elif df.Mileage == 0.0 and df['Brand'] in missing_mileage_brand.index:
            return missing_mileage_brand.at[df['Brand'],'Mileage']
        else:
            return df['Mileage']
    else:
        return df['Mileage']   

# Print the count of missing values in both Train and Test Data set

In [23]:
print('\033[1m'+"Train data set:\n"+'\033[0m' ,train_full_data.isna().sum())
print('\033[1m'+"Test data set:\n"+'\033[0m',test_full_data.isna().sum())

Train data set:
 Name                   0
Location               0
Kilometers_Driven      0
Fuel_Type              0
Transmission           0
Owner_Type             0
Mileage               70
Engine                36
Power                143
Seats                 43
Price                  0
Age                    0
Model                  0
Brand                  0
dtype: int64
Test data set:
 Name                  0
Location              0
Kilometers_Driven     0
Fuel_Type             0
Transmission          0
Owner_Type            0
Mileage              13
Engine               10
Power                32
Seats                11
Age                   0
Model                 0
Brand                 0
dtype: int64


# Fill the missing values in both test and train using the methods created above

# Fill missing values of Power

In [24]:
#Train
train_full_data.Power.fillna(value = 0.0,inplace= True)
train_full_data.Power = train_full_data.apply(fill_na_power,axis =1)

#Test
test_full_data.Power.fillna(value = 0.0,inplace= True)
test_full_data.Power = test_full_data.apply(fill_na_power,axis =1)


# Fill missing values of `Engine`

In [25]:
#Train
train_full_data.Engine.fillna(value = 0.0,inplace= True)
train_full_data.Engine = train_full_data.apply(fill_na_engine,axis =1)

#Test
test_full_data.Engine.fillna(value = 0.0,inplace= True)
test_full_data.Engine = test_full_data.apply(fill_na_engine,axis =1)


# Fill missing values of `Seats`

In [26]:
#Train
train_full_data.Seats.fillna(value = 0.0,inplace= True)
train_full_data.Seats = train_full_data.apply(fill_na_seats,axis =1)

#Test
test_full_data.Seats.fillna(value = 0.0,inplace= True)
test_full_data.Seats = test_full_data.apply(fill_na_seats,axis =1)

# Fill missing values of `Mileage`

In [27]:
#Train
train_full_data.Mileage.fillna(value = 0.0,inplace= True)
train_full_data.Mileage = train_full_data.apply(fill_na_mileage,axis =1)

#Test
test_full_data.Mileage.fillna(value = 0.0,inplace= True)
test_full_data.Mileage = test_full_data.apply(fill_na_mileage,axis =1)

# Still some of the columns have null values i.e 0 values,
# Lets look at those rows

In [28]:
print('Train Mileage:',train_full_data[train_full_data.Mileage == 0.0].Mileage.count())
print('Train Seats:',train_full_data[train_full_data.Seats == 0.0].Seats.count())
print('Train Power:',train_full_data[train_full_data.Power == 0.0].Power.count())
print('Train Engine:',train_full_data[train_full_data.Engine == 0.0].Engine.count())


print('--------------')
print('Test Mileage:',test_full_data[test_full_data.Mileage == 0.0].Mileage.count())
print('Test Seats:',test_full_data[test_full_data.Seats == 0.0].Seats.count())
print('Test Power:',test_full_data[test_full_data.Power == 0.0].Power.count())
print('Test Engine:',test_full_data[test_full_data.Engine == 0.0].Engine.count())

Train Mileage: 1
Train Seats: 0
Train Power: 1
Train Engine: 0
--------------
Test Mileage: 0
Test Seats: 0
Test Power: 0
Test Engine: 0


# Re-ordering the columns

In [29]:
#Re-ordering the columns
train_full_data = train_full_data[['Brand', 'Model', 'Age','Location', 'Kilometers_Driven', 'Fuel_Type', 'Transmission',
      'Owner_Type', 'Mileage', 'Engine', 'Power', 'Seats', 'Price']]
test_full_data = test_full_data[['Brand', 'Model', 'Age','Location', 'Kilometers_Driven', 'Fuel_Type', 'Transmission',
      'Owner_Type', 'Mileage', 'Engine', 'Power', 'Seats']]

# Fiiling the np.nan rows in train data set using imputer

In [30]:
train_full_data.loc[train_full_data[train_full_data.Mileage == 0.0].index,'Mileage'] = np.nan
train_full_data.loc[train_full_data[train_full_data.Power == 0.0].index,'Power'] = np.nan
from sklearn.impute import SimpleImputer
#Training Set Imputation
imputer = SimpleImputer(missing_values = np.nan, strategy = 'most_frequent')
imputer_power= imputer.fit_transform(train_full_data[['Power']]) 
imputer_mileage = imputer.fit_transform(train_full_data[['Mileage']])
train_full_data['Power'] = imputer_power
train_full_data['Mileage'] = imputer_mileage

# Filling the test dataset with the actual car value

In [31]:
test_full_data.loc[test_full_data[test_full_data.Power == 0.0].Power.index,'Power'] = 75

# Again check if the data frame has null values i.e `0.0`

In [32]:
print('Train Mileage:',train_full_data[train_full_data.Mileage == 0.0].Mileage.count())
print('Train Seats:',train_full_data[train_full_data.Seats == 0.0].Seats.count())
print('Train Power:',train_full_data[train_full_data.Power == 0.0].Power.count())
print('Train Engine:',train_full_data[train_full_data.Engine == 0.0].Engine.count())


print('--------------')
print('Test Mileage:',test_full_data[test_full_data.Mileage == 0.0].Mileage.count())
print('Test Seats:',test_full_data[test_full_data.Seats == 0.0].Seats.count())
print('Test Power:',test_full_data[test_full_data.Power == 0.0].Power.count())
print('Test Engine:',test_full_data[test_full_data.Engine == 0.0].Engine.count())

Train Mileage: 0
Train Seats: 0
Train Power: 0
Train Engine: 0
--------------
Test Mileage: 0
Test Seats: 0
Test Power: 0
Test Engine: 0


# Store the DF in a copy for future reference

In [33]:
train_full_data_cp1 = train_full_data.copy()
test_full_data_cp1 =test_full_data.copy()

# Since now we have no missing values , we can proceed with the model building

In [34]:
feature_cols1 = ['Brand', 'Model', 'Age','Location', 'Kilometers_Driven', 'Fuel_Type', 'Transmission','Owner_Type', 'Mileage', 'Engine', 'Power', 'Seats']

# Label encoding for categorical variables

In [35]:
#Initializing label encoders
from sklearn.preprocessing import LabelEncoder
le_brands = LabelEncoder()
le_models = LabelEncoder()
le_locations = LabelEncoder()
le_fuel_types = LabelEncoder()
le_transmissions = LabelEncoder()
le_owner_types = LabelEncoder()

In [36]:
all_brands = list(set(list(train_full_data.Brand) + list(test_full_data.Brand)))
all_models = list(set(list(train_full_data.Model) + list(test_full_data.Model)))
all_locations = list(set(list(train_full_data.Location) + list(test_full_data.Location)))
all_fuel_types = list(set(list(train_full_data.Fuel_Type) + list(test_full_data.Fuel_Type)))
all_transmissions = list(set(list(train_full_data.Transmission) + list(test_full_data.Transmission)))
all_owner_types = list(set(list(train_full_data.Owner_Type) + list(test_full_data.Owner_Type)))

In [37]:
#Fitting the categories
le_brands.fit(all_brands)
le_models.fit(all_models)
le_locations.fit(all_locations)
le_fuel_types.fit(all_fuel_types)
le_transmissions.fit(all_transmissions)
le_owner_types.fit(all_owner_types)

LabelEncoder()

In [38]:
#Applying encoding to test_full_data data
test_full_data['Brand'] = le_brands.transform(test_full_data['Brand'])
test_full_data['Model'] = le_models.transform(test_full_data['Model'])
test_full_data['Location'] = le_locations.transform(test_full_data['Location'])
test_full_data['Fuel_Type'] = le_fuel_types.transform(test_full_data['Fuel_Type'])
test_full_data['Transmission'] = le_transmissions.transform(test_full_data['Transmission'])
test_full_data['Owner_Type'] = le_owner_types.transform(test_full_data['Owner_Type'])

In [39]:
#Applying encoding to train_full_data data
train_full_data['Brand'] = le_brands.transform(train_full_data['Brand'])
train_full_data['Model'] = le_models.transform(train_full_data['Model'])
train_full_data['Location'] = le_locations.transform(train_full_data['Location'])
train_full_data['Fuel_Type'] = le_fuel_types.transform(train_full_data['Fuel_Type'])
train_full_data['Transmission'] = le_transmissions.transform(train_full_data['Transmission'])
train_full_data['Owner_Type'] = le_owner_types.transform(train_full_data['Owner_Type'])

In [40]:
train_full_data.head()

Brand  Model  Age  Location  Kilometers_Driven  Fuel_Type  Transmission  \
0     18    612    9         9              72000          0             1   
1     11    162    4        10              41000          1             1   
2     10    335    8         2              46000          4             1   
3     18    224    7         2              87000          1             1   
4      1     21    6         3              40670          1             0   

   Owner_Type  Mileage  Engine   Power  Seats  Price  
0           0    26.60   998.0   58.16    5.0   1.75  
1           0    19.67  1582.0  126.20    5.0  12.50  
2           0    18.20  1199.0   88.70    5.0   4.50  
3           0    20.77  1248.0   88.76    7.0   6.00  
4           2    15.20  1968.0  140.80    5.0  17.74

In [41]:
#Train
X_train_data = train_full_data.loc[:,train_full_data.columns != 'Price']
y_train_data = train_full_data['Price']

#test
X_test_data = test_full_data

X_test_data.head()

Brand  Model  Age  Location  Kilometers_Driven  Fuel_Type  Transmission  \
0     18      1    5         4              40929          0             1   
1     18      1    6         3              54493          4             1   
2     29      1    2         9              34000          1             1   
3     29      1    7         5             139000          1             1   
4     11      1    5         9              29000          4             1   

   Owner_Type  Mileage  Engine   Power  Seats  
0           0    32.26   998.0   58.20    4.0  
1           2    24.70   796.0   47.30    5.0  
2           0    13.68  2393.0  147.80    7.0  
3           0    23.59  1364.0   90.00    5.0  
4           0    18.50  1197.0   82.85    5.0

In [42]:
from sklearn.model_selection import train_test_split

X_train, X_val, Y_train, Y_val = train_test_split(X_train_data, y_train_data, test_size = 0.2, random_state = 1)

In [43]:
from sklearn.metrics import mean_absolute_error
from sklearn import metrics
# Score Calculation
def score(y_pred, y_true):
   error = np.square(np.log10(y_pred +1) - np.log10(y_true +1)).mean() ** 0.5
   score = 1 - error
   return score

In [44]:
from sklearn.ensemble import RandomForestRegressor
model_rfr = RandomForestRegressor(n_estimators=90 ,random_state=0)
model_rfr.fit(X_train,Y_train)
y_pred = model_rfr.predict(X_val)
print(mean_absolute_error(Y_val,y_pred))
#Printing the score for validation sets
print("\n\n  RandomForestRegressor SCORE : ", score(y_pred, Y_val))

1.4507648879469965


  RandomForestRegressor SCORE :  0.9288926275577388


In [45]:
from sklearn.ensemble import RandomForestRegressor
model_rfr = RandomForestRegressor(n_estimators=90 ,random_state=0)
model_rfr.fit(X_train_data,y_train_data)
y_pred = model_rfr.predict(X_test_data)

In [46]:
#Saving the predictions to an excel sheet
pd.DataFrame(y_pred, columns = ['Price']).to_excel("predictions.xlsx")